bpfd/src/bpf.rs

```diff
@@ -875,66 +868,71 @@ impl BpfManager {
     // program is the map owner, then a new entry is add to the map hash
     // table and permissions on the directory are updated to grant bpfd
     // user group access to all the maps in the directory. If this eBPF
-    // program is not the owner, then the eBPF program UUID is added to
+    // program is not the owner, then the eBPF program ID is added to
     // the Used-By array.
     async fn save_map(
         &mut self,
-        id: Uuid,
-        map_owner_uuid: Option<Uuid>,
+        id: u32,
+        map_owner_id: Option<u32>,
         map_pin_path: &Path,
     ) -> Result<(), BpfdError> {
-        let (map_owner, _) = get_map_index(id, map_owner_uuid);
-
-        if map_owner {
-            let program = self
-                .programs
-                .get_mut(&id)
-                .expect("Program should be loaded");
-
-            let map = BpfMap { used_by: vec![id] };
-
-            self.maps.insert(id, map);
+        match map_owner_id {
+            Some(id) => {
+                if let Some(map) = self.maps.get_mut(&id) {
+                    map.used_by.push(id);
+
+                    // If map owner program still exists update it's maps_used_by_field
+                    if let Some(program) = self.programs.get_mut(&id) {
```

Comment:
Now `program.data.used_by` is what is returned in list/get, so this `if` block needs to be in a loop, and all programs with an instance in this `map.used_by` need to be updated with this new `id`.

The same will need to be done when a program in the `map.used_by` is unloaded. Also, `program.data.used_by` is not being updated on a bpfd restart.

The problem is you are trying to keep the same data in two different places and now you have to keep them in sync. And its a many `program.data.used_by` to one `map.used_by`. It's probably best to not keep `used_by` in ProgramData.